File: problems/279/279.perfect-squares.py
#
# @lc app=leetcode id=279 lang=python3
#
# [279] Perfect Squares
#

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n] * (n + 1)
        dp[0] = 0
        squares = [x**2 for x in range(0, n) if x**2 <= n]

        for target in range(1, n + 1):
            for square in squares:
                if target - square < 0:
                    break
                dp[target] = min(dp[target], 1 + dp[target - square])
        
        return dp[n]
        
# @lc code=end

File: problems/279/279.perfect-squares.java
/*
 * @lc app=leetcode id=279 lang=java
 *
 * [279] Perfect Squares
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int numSquares(int n) {
        /*
         * given n, return the least number of perfect square numbers that sum to n.
         * perfect squares -> 1,4,9,16,.. (integer that is the square of an integer)
         *
         * intuition:
         * - brute force bottom up dp
         * - dp[i] -> min number of perfect squares to make value i
         *
         * O(N * sqrt(N)) time, O(N) space.
         */

        int[] dp = new int[n + 1];
        Arrays.fill(dp, n); // max number of vals it takes is 1 * n
        // base cases
        dp[0] = 0;
        dp[1] = 1;

        for (int target = 2; target <= n; target++) {
            for (int curr = 1; curr <= Math.floor(Math.sqrt(target)); curr++) {
                int sqr = curr * curr;
                if (target - sqr < 0) {
                    break;
                }

                dp[target] = Math.min(dp[target], dp[target - sqr] + 1);
            }
        }

        return dp[n];
    }
}
// @lc code=end
